### PR TITLE
Add unused-args warning flags

### PR DESCRIPTION
**Pull request by richsalz**
This adds -Wunused-param and fixes many warnings.  Not all; still over 600 to go.

Is this worth doing?  Most of the warnings are because of having to fit into a common API.  But not all, e.g., DSA_sign() could have been changed.

Thoughts?

### CODE REVIEW

**Comment by levitte** [comment]
I've a thought, change lines like this:

``` C
    osslargused(foo); osslargused(bar);
```

to:

``` C
    osslargused(foo);
    osslargused(bar);
```

**Comment by richsalz** [comment]
Okay, @levitte , update commit pushed which does that (one argsused per line)

**Comment by levitte** [comment]
Generally speaking, I like the idea.

**Comment by dot-asm** [comment]
This might be a bit off-topic, but since request was quite general "thoughts?" I decided to say couple of things as a *side note*.

I for one would find straightforward `(void)x` more appropriate. Because more "raw" C is better for readability, as well as fosters "portability" discipline. I mean I see lesser value in inventing kind of a lingo unless it saves a lot of space. Macro for `(void)x` doesn't save any space at all. But this is not as important as following.

I find placement of the macro unsatisfactory. e_os.h is quite "heavy" file, and justifying its inclusion for sole purpose of macro as trivial as `(void)x` is quite an overkill. Yes, one can argue that e_os.h included a lot, so that couple of more inclusions wouldn't make difference, but I also said that it might be a bit off-topic, didn't I :-) Consider aes_ige.c for example. Does one actually need to include socket.h? tcp.h? windows.h? vxWorks.h? No, all it needs from system headers is string.h.

I suppose bottom line is that we have a lot of code that appears [or could appear] quite generic to *a* C programmer, and I'd argue that we should strive for making as much code as possible to be that way, not vice versa. And minimizing dependencies from system headers is a good thing...

**Comment by levitte** [comment]
I agree that the overuse of e_os.h isn't the best choice. We have the ability for internal headers, would one for these simple macros be possible? 

I disagree regarding the "purity of language" line of thought... Spurious `(void)x` with no explanation whatsoever will only look strange a few years ahead, and even more so to anyone entering the scene later on.

**Comment by dot-asm** [comment]
> I disagree regarding the "purity of language" line of thought...

In general or in this particular context?

> Spurious (void)x with no explanation whatsoever will only look strange a few years ahead, and even more so to anyone entering the scene later on.

I'd say quite contrary. Made-up macros look (mote present tense) strange to anyone entering the scene at any given point. And I'd say they would be better off learning the universal idiom than to look for project-specific lingvo.

BTW, out of curiosity how would you find following:
```
int foo(int i, int j)
{
    return i;
    (void)j;
}
```

**Comment by richsalz** [comment]
The issue is that on some compilers ```(void)x``` creates a warning, as I recall.  So osslargused iseems to be more portable.

**Comment by richsalz** [comment]
you guys are right, e_os.h is too heavyweight.  Perhaps I should refactor it?  Got any clues for filenames?

**Comment by levitte** [comment]
> The issue is that on some compilers ```(void)x``` creates a warning, as I recall.  So osslargused iseems to be more portable.

I fail to see how a macro that translates `x` into `(void)x` would avoid that issue...

**Comment by levitte** [comment]
@dot-asm said thusly:
> > I disagree regarding the "purity of language" line of thought...
> 
> In general or in this particular context?

I was mostly thinking in this particular context...  but thinking of it, I think that spurious no-ops like `(void)x;` will raise eye-brows.

> > Spurious (void)x with no explanation whatsoever will only look strange a few years ahead, and even more so to anyone entering the scene later on.
> 
> I'd say quite contrary. Made-up macros look (mote present tense) strange to anyone entering the scene at any given point. And I'd say they would be better off learning the universal idiom than to look for project-specific lingvo.

Since when is `(void)x` a universal idiom for "we want to shut the compiler up"?

> BTW, out of curiosity how would you find following:
> ```
> int foo(int i, int j)
> {
>     return i;
>     (void)j;
> }
> ```

Weird.  That's how I'd find it.  And I'd ask around.  If, however, I found something like `osslargused(x);` (or `protect(x);`, something I've seen elsewhere quite a while ago), I'd look around the headers to find out and see if I can find the documentation (I just noticed that it's not documented at all...  @richsalz, that needs fixing!).

Side side note: osslargused is already defined since last spring, and used in `crypto/mem.c`...

**Comment by dot-asm** [comment]
> Since when is (void)x a universal idiom for "we want to shut the compiler up"?

Well, what I'm saying is that we would be better off is there was one *and* that straightforward `(void)x` is the only viable candidate really(`*`). Yes, it might appear eye-browsing for the first time, but then reader very soon realize what it's about and easily recognize the pattern. But if each project has own quirk, then it becomes troublesome to switch, not to mention that formally reader would be compelled [if not obliged] to verify if it's still same since last time.

(`*`) Of course there are compiler-specific extensions to mark unused variables, but we don't want to go there, do we?

**Comment by dot-asm** [comment]
> Weird. That's how I'd find it. And I'd ask around.

Well, that was kind of exaggeration (placing `(void)unsused_var` past return) and was meant rather as "shake up your mind" than actual suggestion. It would be fun to hear what others think of having their mind shaken that way, but you don't have to ask...

> Side side note: osslargused is already defined since last spring, and used in crypto/mem.c...

See Tomas Mraz's signature :-) Please do note smiley and recall original assertion that it might be off-topic.

File: crypto/ec/ec2_smpl.c [hunk 1]
```diff
@@ -389,6 +395,7 @@ int ec_GF2m_simple_point_get_affine_coordinates(const EC_GROUP *group,
 {
     int ret = 0;
 
+    osslargused(tx);
```
Comment (levitte) [inline crypto/ec/ec2_smpl.c]: s/tx/ctx/

**Comment by levitte** [comment]
> you guys are right, e_os.h is too heavyweight.  Perhaps I should refactor it?  Got any clues for filenames?

`include/internal/util.h`  (for "utility macros and whatnot")

**Comment by richsalz** [comment]
for platforms that complain about cast-to-avoid, we make osslargused be a no-op

**Comment by levitte** [comment]
> for platforms that complain about cast-to-avoid, we make osslargused be a no-op

Ah, yes.  And that's definitely a reason for a macro...

File: crypto/ec/ec2_smpl.c (hunk 1 above)
Comment (richsalz) [inline crypto/ec/ec2_smpl.c]: was fixed, will be in next push tnx
:)

Comment (levitte) [inline crypto/ec/ec2_smpl.c]: Okie

**Comment by dot-asm** [comment]
>>    for platforms that complain about cast-to-avoid, we make osslargused be a no-op
>
> Ah, yes. And that's definitely a reason for a macro...

If such platform can be identified. But even if one would be found, we still can choose to ignore the warning (it has to be warning as cast doesn't violate standard). We don't have to work around every non-compliant quirk...

**Comment by richsalz** [comment]
I'm fine with the void cast if the team thinks it's more clear.  The larger question still remains -- is this worth doing?

**Comment by petrovr** [comment]
Using macro ...UNUSED...  defined as (void)arg is quite common way to suppress compiler warning for unused function arguments.
You could find macro defined in lot of projects . The difference is that macro is in uppercase and has prefix UN(USED).
If consensus is to refactor and use it please use UNUSED as part of name.

**Comment by levitte** [comment]
I guess that answer the question on idioms ;-)

**Comment by richsalz** [comment]
Well it's one person's view.  C++ allows nameless params, which is great but we can't use it.  Others use "USED" to indicate to tell the compiler that the argument is "used" .

Again, the main question is: do this or not?

**Comment by dot-asm** [comment]
> I guess that answer the question on idioms ;-)

In my mind it rather illuminates my point. That every project has to make up own macro name, and my point is that it doesn't *really* have to be that way. Yes, macro's author would argue that it's perfectly readable in every way, but is it *really* better than straightforward `(void)x` for *everybody*? How about following? Unused variable should be *named* as such, e.g. nm_unused, and then can be idiom-ed with `(void)`.

To answer the question if it's worth doing, Rich. Richard says yes, I'd also say yes. How exactly it's done is presumably up to you. I mean Richard is pro-macro, I'm against, so *you* get to decide :-) We all seem to agree that placement of the macro is less optimal, i.e. *if* it will be macro after all...

**Comment by richsalz** [comment]
I'm gonna go with the void cast directly.  I'll work on this next week.  For now, just updating this PR shortly about the void and new header file (not needed) thing.

**Comment by dot-asm** [comment]
> I'm gonna go with the void cast directly.

Do consider renaming even variables, as hint for why cast is there. But not in declarations, in implementations...

**Comment by richsalz** [comment]
I did think of it, calling them 'unused1' etc.  but i think the decl and the implementation should match exactly.  and also too much work ...

**Comment by dot-asm** [comment]
> the decl and the implementation should match exactly.

Types have to match, not argument names, no.

**Comment by richsalz** [comment]
I know the names don't have to match.  :)  But when they don't it jangles and makes me think something is not right.

**Comment by richsalz** [comment]
New single unified commit pushed.  one-third finished.  Will wait to see if anyone objects before I finish this off.

**Comment by levitte** [comment]
I just have to try this on VMS ;-)

**Comment by dot-asm** [comment]
I get warnings in number of files in crypto/x509v3, engines/e_padlock.c, few files in ssl/, ssl/statem (strange, those are relatively new). Then it fails to compile apps/apps.c:app_init, s/msgwin/mesgwin/ in line#121, warnings in apps/. I mean that's what CI is going to have to say. VMS should fail to compile apps/apps.c...

**Comment by richsalz** [comment]
it's not done yet :)

**Comment by dot-asm** [comment]
It would probably be appropriate to mention this in coding style... I'd also argue that last sentence in chapter 6 could be relaxed to permit _usused suffix or unused_ prefix in function implementation. Though there is bunch of cases where additional pre-/suffix doesn't formally conflict with the last sentence of the chapter 6. Cases when *static* methods get implemented and don't require all the arguments...

**Comment by richsalz** [comment]
okay, complete.  some build variants will fail, but let's see how travis does.  then i'll run check-build.

**Comment by dot-asm** [comment]
It complains about argc/argv in three test/programs.

**Comment by richsalz** [comment]
which ones?  I have to do a run-checker set of tests now to flush these things out.  But what config flags did you use and where did it complain?

**Comment by dot-asm** [comment]
memleaktest.c, rc5test.c, wpackettest.c, clang [3.9] configured with --strict-warnings.

File: apps/s_server.c [hunk 2]
```diff
@@ -147,13 +147,14 @@ static int dtlslisten = 0;
 static char *psk_identity = "Client_identity";
 char *psk_key = NULL;           /* by default PSK is not used */
 
-static unsigned int psk_server_cb(SSL *ssl, const char *identity,
+static unsigned int psk_server_cb(SSL *s, const char *identity,
```
Comment (dot-asm) [inline apps/s_server.c]: You are being inconsistent with what you said/implied before :-) You said that your preferences are with argument *names* in declarations and definitions matching exactly. And there is support for this in style guide. Well, specifically here there is a level of indirection, i.e. it is the pointer to this function that is used, but I'd expect you to express your preferences even here :-) As already mentioned earlier, myself I'd argue that argument name matching is not in contradiction with style guide in cases of indirection like above, so I'd accept it. Though I'd say that if renamed `ssl_unused` would be a 100 times better name in this case. But that's just me.

BTW, *if* style guide is amended to permit unused pre-/suffix, wouldn't it be more optimal to harmonize with it here at once? I mean as opposite to committing this now and then going through it once again later on?

Comment (richsalz) [inline apps/s_server.c]: yes, there are a few places where habits kicked in and I did some other code-cleanup.  Working on backing all of those out now.

perhaps you're right about the unused thing.  i'll think about it.

Comment (dot-asm) [inline apps/s_server.c]: Just to clarify. It should be obvious from what was said yesterday, but "argument name matching is not in contradiction with style guide in cases of indirection like above" came out wrong, as I meant "argument *renaming* is not in contradiction [in cases of *indirection*]".

**Comment by richsalz** [comment]
I understood you.  And I pulled all the other style changes out, so a diff against master shows only void casts.  (With one exception, an indent fix :)

**Comment by richsalz** [comment]
Almost every option builds now.  I'm fixing cases as I find them.  @dot-asm, please take a look at the latest commit.  Sometimes I put the void casts at the end. This is because of the common case of a function being completely ifdef'd out.  I could reverse the test, do the void casts, and then have the real function body.  Should I?

**Comment by levitte** [comment]
Please add `[skip ci]` at the end of the last commit while you're doing that work.  That will avoid queueing up that huge amount of Travis builds (a new one for each push).

**Comment by levitte** [comment]
Quite frankly, the [dispute about number of arguments in 2341](https://github.com/openssl/openssl/pull/2341#discussion_r99363518) helped me form a stronger opinion about this effort.

As this effort seems to become a hindrance of expression more than a tool to find actual faults, I cannot see that it's worth the while.  I can't myself approving this, sorry.

**Comment by richsalz** [comment]
You got it backwards.  If there were a struct with parameters in it, then none of the void casts would be necessary.

**Comment by levitte** [comment]
Errr, I think you misunderstood what I said

**Comment by richsalz** [comment]
I don't think I did.  You think it's okay to have many parameters even if some percentage aren't used.  And that therefore the unused-arg flag is not worthwhile.  I disagree.  We can put things in a struct, there are no unused args in all those functions, and the unused-arg flag becomes worthwhile.  I also think long parameter lists hinder readability and maintainability.

**Comment by levitte** [comment]
For the sake of the argument, I'll copy a comment from #2341:

> I both agree and disagree with you.  I agree *if* it's a tool to secure a stable (and a priori public) API while keeping the possible input to affected function flexible, that avoids having a gazillion different functions, one for each possible case.  For internal functions, I don't see the point unless there's a *huge* (6 isn't huge) number of parameters that are scarcelly used.

In other words, I think that making a very strict choice toward custom structs any time there might be some unused parameters is exagerated.  However, I do see some value in it, but would prefer determining that on a case by case basis.

And mind you, my darling operating system VMS has a fairly large number of library routines that work with the kind custom structs you talk about...  in their *public* API.  For a small example if what that looks like, see `crypto/LPdir_vms.c`.

**Comment by richsalz** [comment]
I don't believe I've ever said a strict decision.  I *have* said that we've already seen this grow, and that six is the tipping point (for me).   This is all about parsing extensions, put the parameters into a sub-struct in the SSL object.  That's where all other state is ...

**Comment by levitte** [comment]
You keep refering to he SSL structure.  Are there other places where we put values in the SSL structure *just to be able to pass parameters*, or do we do it to *cache the values* to be used elsewhere, not just in the next function we call?  That's a *crucial* difference in my mind.

**Comment by richsalz** [comment]
We do both.  But I don't care all that much if it's "SSL *s, TLSEXTPARSESTATE *st"" two parameters or the st is inside the s.  I do believe that six (and we're not done yet) is unwieldly, hard to read, obscures unused-arg errors that some would like to see, and too rigid for future expansion.

**Comment by richsalz** [comment]
closing for lack of support within the team.

**Pull request closed without merging**